File: examples/plastic-resin-price-scraper/node/scrapeResinPrices.mjs
/**
 * Cao gia nguyen vat lieu nhua nguyen sinh PP / PE bang Firecrawl Cloud (Node.js).
 *
 * Cach dung:
 *   export FIRECRAWL_API_KEY=fc-...
 *   npm install
 *   npm start
 *
 * Ket qua ghi ra thu muc ../output/ duoi dang JSON + CSV.
 * Cau hinh nguon va schema o file ../sources.json.
 */

import { readFile, mkdir, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import Firecrawl from "@mendable/firecrawl-js";

const HERE = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(HERE, "..");
const CONFIG_PATH = resolve(ROOT, "sources.json");
const OUTPUT_DIR = resolve(ROOT, "output");

const CSV_FIELDS = [
  "source",
  "material",
  "grade",
  "price",
  "currency",
  "unit",
  "region",
  "date",
  "change",
  "scraped_at",
];

function fail(msg) {
  console.error(msg);
  process.exit(1);
}

async function loadConfig() {
  try {
    return JSON.parse(await readFile(CONFIG_PATH, "utf-8"));
  } catch (err) {
    fail(`Khong doc duoc cau hinh ${CONFIG_PATH}: ${err.message}`);
  }
}

function buildClient() {
  const apiKey = process.env.FIRECRAWL_API_KEY;
  if (!apiKey) {
    fail(
      "Thieu FIRECRAWL_API_KEY. Lay key tai https://firecrawl.dev roi:\n" +
        "    export FIRECRAWL_API_KEY=fc-...",
    );
  }
  return new Firecrawl({ apiKey });
}

async function scrapeSource(client, source, schema, scrapeCfg) {
  const { name, url } = source;
  console.log(`  -> [${name}] dang scrape: ${url}`);

  const jsonFormat = { type: "json", schema };
  if (source.prompt) jsonFormat.prompt = source.prompt;

  let doc;
  try {
    doc = await client.scrape(url, {
      formats: [jsonFormat],
      onlyMainContent: scrapeCfg.onlyMainContent ?? true,
      timeout: scrapeCfg.timeoutMs ?? 60000,
      maxAge: scrapeCfg.maxAgeMs ?? 3600000,
    });
  } catch (err) {
    console.log(`     [LOI] ${name}: ${err.message}`);
    return [];
  }

  const data = doc?.json ?? {};
  const prices = Array.isArray(data.prices) ? data.prices : [];
  const scrapedAt = new Date().toISOString();

  const rows = prices
    .filter((p) => p && typeof p === "object")
    .map((p) => ({
      source: name,
      material: p.material ?? source.material ?? "",
      grade: p.grade ?? "",
      price: p.price ?? "",
      currency: p.currency ?? "",
      unit: p.unit ?? "",
      region: p.region ?? "",
      date: p.date ?? "",
      change: p.change ?? "",
      scraped_at: scrapedAt,
    }));

  console.log(`     OK: ${rows.length} dong gia`);
  return rows;
}

function toCsv(rows) {
  const escape = (v) => {
    const s = String(v ?? "");
    return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
  };
  const header = CSV_FIELDS.join(",");
  const lines = rows.map((r) => CSV_FIELDS.map((f) => escape(r[f])).join(","));
  return [header, ...lines].join("\n");
}

async function writeOutputs(rows) {
  await mkdir(OUTPUT_DIR, { recursive: true });
  const stamp = new Date()
    .toISOString()
    .replace(/[:T]/g, "")
    .replace(/\..+/, "")
    .replace(/-/g, "");
  const jsonPath = resolve(OUTPUT_DIR, `resin_prices_${stamp}.json`);
  const csvPath = resolve(OUTPUT_DIR, `resin_prices_${stamp}.csv`);

  await writeFile(jsonPath, JSON.stringify(rows, null, 2), "utf-8");
  await writeFile(csvPath, toCsv(rows), "utf-8");

  console.log(`\nDa ghi ${rows.length} dong gia vao:`);
  console.log(`  - ${jsonPath}`);
  console.log(`  - ${csvPath}`);
}

async function main() {
  const cfg = await loadConfig();
  const client = buildClient();
  const schema = cfg.schema;
  const scrapeCfg = cfg.scrape ?? {};
  const sources = (cfg.sources ?? []).filter((s) => s.enabled);

  if (sources.length === 0) {
    fail("Khong co nguon nao duoc bat (enabled=true) trong sources.json");
  }

  console.log(`Bat dau cao gia tu ${sources.length} nguon...\n`);
  const allRows = [];
  for (const source of sources) {
    allRows.push(...(await scrapeSource(client, source, schema, scrapeCfg)));
  }

  if (allRows.length === 0) {
    console.log("\nKhong trich xuat duoc dong gia nao. Kiem tra lai URL / prompt / schema.");
    return;
  }

  await writeOutputs(allRows);
}

main().catch((err) => fail(err.stack || String(err)));

File: examples/plastic-resin-price-scraper/python/scrape_resin_prices.py
"""
Cao gia nguyen vat lieu nhua nguyen sinh PP / PE bang Firecrawl Cloud.

Cach dung:
    export FIRECRAWL_API_KEY=fc-...
    pip install -r requirements.txt
    python scrape_resin_prices.py

Ket qua ghi ra thu muc ../output/ duoi dang JSON + CSV.

Cau hinh nguon va schema o file ../sources.json (sua url / enabled / prompt tuy y).
"""

import csv
import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

from firecrawl import Firecrawl

HERE = Path(__file__).resolve().parent
ROOT = HERE.parent
CONFIG_PATH = ROOT / "sources.json"
OUTPUT_DIR = ROOT / "output"

# Cac cot xuat ra CSV (theo schema trong sources.json)
CSV_FIELDS = [
    "source",
    "material",
    "grade",
    "price",
    "currency",
    "unit",
    "region",
    "date",
    "change",
    "scraped_at",
]


def load_config() -> dict:
    if not CONFIG_PATH.exists():
        sys.exit(f"Khong tim thay file cau hinh: {CONFIG_PATH}")
    with CONFIG_PATH.open("r", encoding="utf-8") as f:
        return json.load(f)


def build_client() -> Firecrawl:
    api_key = os.environ.get("FIRECRAWL_API_KEY")
    if not api_key:
        sys.exit(
            "Thieu FIRECRAWL_API_KEY. Lay key tai https://firecrawl.dev roi:\n"
            "    export FIRECRAWL_API_KEY=fc-..."
        )
    return Firecrawl(api_key=api_key)


def scrape_source(client: Firecrawl, source: dict, schema: dict, scrape_cfg: dict) -> list[dict]:
    """Scrape mot nguon, tra ve danh sach dong gia da lam phang."""
    name = source["name"]
    url = source["url"]
    print(f"  -> [{name}] dang scrape: {url}")

    json_format = {"type": "json", "schema": schema}
    if source.get("prompt"):
        json_format["prompt"] = source["prompt"]

    try:
        doc = client.scrape(
            url,
            formats=[json_format],
            only_main_content=scrape_cfg.get("onlyMainContent", True),
            timeout=scrape_cfg.get("timeoutMs", 60000),
            max_age=scrape_cfg.get("maxAgeMs", 3600000),
        )
    except Exception as exc:  # noqa: BLE001 - bao loi tung nguon, khong dung ca chuong trinh
        print(f"     [LOI] {name}: {exc}")
        return []

    data = getattr(doc, "json", None) or {}
    prices = data.get("prices", []) if isinstance(data, dict) else []
    scraped_at = datetime.now(timezone.utc).isoformat()

    rows = []
    for p in prices:
        if not isinstance(p, dict):
            continue
        rows.append(
            {
                "source": name,
                "material": p.get("material") or source.get("material", ""),
                "grade": p.get("grade", ""),
                "price": p.get("price", ""),
                "currency": p.get("currency", ""),
                "unit": p.get("unit", ""),
                "region": p.get("region", ""),
                "date": p.get("date", ""),
                "change": p.get("change", ""),
                "scraped_at": scraped_at,
            }
        )
    print(f"     OK: {len(rows)} dong gia")
    return rows


def write_outputs(rows: list[dict]) -> None:
    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
    stamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    json_path = OUTPUT_DIR / f"resin_prices_{stamp}.json"
    csv_path = OUTPUT_DIR / f"resin_prices_{stamp}.csv"

    with json_path.open("w", encoding="utf-8") as f:
        json.dump(rows, f, ensure_ascii=False, indent=2)

    with csv_path.open("w", encoding="utf-8", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=CSV_FIELDS)
        writer.writeheader()
        writer.writerows(rows)

    print(f"\nDa ghi {len(rows)} dong gia vao:")
    print(f"  - {json_path}")
    print(f"  - {csv_path}")


def main() -> None:
    cfg = load_config()
    client = build_client()
    schema = cfg["schema"]
    scrape_cfg = cfg.get("scrape", {})
    sources = [s for s in cfg["sources"] if s.get("enabled", False)]

    if not sources:
        sys.exit("Khong co nguon nao duoc bat (enabled=true) trong sources.json")

    print(f"Bat dau cao gia tu {len(sources)} nguon...\n")
    all_rows: list[dict] = []
    for source in sources:
        all_rows.extend(scrape_source(client, source, schema, scrape_cfg))

    if not all_rows:
        print("\nKhong trich xuat duoc dong gia nao. Kiem tra lai URL / prompt / schema.")
        return

    write_outputs(all_rows)


if __name__ == "__main__":
    main()
